zcash_test_vectors/zip_0032.py

```diff
@@ -53,12 +58,62 @@ def master(cls, ContextString, S):
         assert 32 <= length_S <= 252
 
         IKM = bytes([length_ContextString]) + ContextString + bytes([length_S]) + S
-        (sk, chaincode) = MKGh(cls.Arbitrary, IKM)
+        (sk, chaincode) = MKGh(cls.Context, IKM)
         return cls(IKM, [], sk, chaincode)
 
     def child(self, i):
-        (sk_i, c_i) = CKDh(self.Arbitrary, self.sk, self.chaincode, i)
-        return self.__class__(None, self.path + [i], sk_i, c_i)
+        (sk_child, c_child) = CKDh(self.Context, self.sk, self.chaincode, i, b"", False)
+        return self.__class__(None, self.path + [i], sk_child, c_child)
+
+    def child_with_tag(self, i, tag):
+        (sk_child, c_child) = CKDh(self.Context, self.sk, self.chaincode, i, tag, False)
+        (left, right) = CKDh(self.Context, self.sk, self.chaincode, i, tag, True)
+        return self.__class__(None, self.path + [(i, tag)], sk_child, c_child, left + right)
+
+
+def registered_key_derivation_tvs():
+    args = render_args()
+
+    context_string = b'Zcash test vectors'
+    seed = bytes(range(32))
+    m = RegisteredKey.master(context_string, seed)
+    m_1h = m.child(hardened(1))
+    m_1h_2h = m_1h.child_with_tag(hardened(2), b"trans rights are human rights")
+    m_1h_2h_3h = m_1h_2h.child_with_tag(hardened(3), b"")
+
+    keys = [m_1h, m_1h_2h, m_1h_2h_3h]
+
+    test_vectors = [
+        {
+            'context_string': context_string,
+            'seed':       seed,
+            'zip_number': 1,
+            'subpath':    k.path[1:],
+            'sk':         k.sk,
+            'c':          k.chaincode,
+            'full_width': k.full_width,
+        }
+        for k in keys
+    ]
+
+    render_tv(
+        args,
+        'zip_0032_registered',
+        (
+            ('context_string', '&\'static [u8]'),
+            ('seed',       '[u8; 32]'),
+            ('zip_number', 'u16'),
+            ('subpath',    '&\'static [(u32, &\'static [u8])]'),
+            ('sk',         '[u8; 32]'),
+            ('c',          '[u8; 32]'),
+            ('full_width', 'Option<[u8; 64]>'),
+        ),
+        test_vectors,
+    )
+
+
+class ArbitraryKey(RegisteredKey):
```

Comment:
This should be reverted; ZIP 32 does not have ArbitraryKey as a sub-instance of RegisteredKey, and the intention of the test vector implementations is to be as close to the specifications as possible (vs the production impls which may take liberties with APIs for performance reasons).